p2panda-rs/src/schema/schema_id.rs

```diff
@@ -174,37 +204,70 @@ mod test {
         );
     }
 
-    #[test]
-    fn invalid_deserialization() {
-        assert!(serde_json::from_str::<SchemaId>("[\"This is not a hash\"]").is_err());
-        assert!(serde_json::from_str::<SchemaId>("5").is_err());
-        assert!(serde_json::from_str::<SchemaId>(
-            "0020c65567ae37efea293e34a9c7d13f8f2bf23dbdc3b5c7b9ab46293111c48fc78b"
-        )
-        .is_err());
-
-        // Test invalid hash
-        let invalid_hash = serde_json::from_str::<SchemaId>(
-            "[\"0020c65567ae37efea293e34a9c7d13f8f2bf23dbdc3b5c7b9ab46293111c48fc7\"]",
-        );
+    // Not a hash at all
+    #[rstest]
+    #[case(
+        "\"This is not a hash\"",
+        "malformed application schema id: expecting name and view id hashes separated by \
+        underscore at line 1 column 20"
+    )]
+    // An integer
+    #[case(
+        "5",
+        "invalid type: integer `5`, expected schema id as string at line 1 column 1"
+    )]
+    // Only an operation id, could be interpreted as document view id but still missing the name
+    #[case(
+        "\"0020c65567ae37efea293e34a9c7d13f8f2bf23dbdc3b5c7b9ab46293111c48fc78b\"",
+        "malformed application schema id: expecting name and view id hashes separated by \
+        underscore at line 1 column 70"
+    )]
+    // Only the name is missing now
+    #[case(
+        "\"_0020c65567ae37efea293e34a9c7d13f8f2bf23dbdc3b5c7b9ab46293111c48fc78b\"",
+        "application schema id is missing a name: _0020c65567ae37efea293e34a9c7d13f8f2bf23dbdc3b5c\
+        7b9ab46293111c48fc78b at line 1 column 71"
+    )]
+    // This name is too long, parser will fail trying to read its last section as an operation id
+    #[case(
+        "\"this_name_is_way_too_long_it_cant_be_good_to_have_such_a_long_name_to_be_honest_0020c65\
+        567ae37efea293e34a9c7d13f8f2bf23dbdc3b5c7b9ab46293111c48fc78b\"",
+        "encountered invalid hash while parsing application schema id: invalid hex encoding in \
+        hash string at line 1 column 150"
+    )]
+    // This hash is malformed
+    #[case(
+        "\"venue_0020c65567ae37efea293e34a9c7d13f8f2bf23dbdc3b5c7b9ab46293111c48fc7\"",
+        "encountered invalid hash while parsing application schema id: invalid hash length 33 \
+    bytes, expected 34 bytes at line 1 column 74"
+    )]
+    // this looks like a system schema, but it is not
+    #[case(
+        "\"unknown_system_schema_name_v1\"",
+        "not a known system schema: unknown_system_schema_name_v1 at line 1 column 31"
+    )]
+    fn invalid_deserialization2(#[case] schema_id: &str, #[case] expected_err: &str) {
```

Comment:
Really nice use of rstest cases!! I'd not used it like this so much so far (prefering the templates) because I felt it got hard to read, but this works so well with just strings!

Reply:
Automatic code formatting is not working for them which is a bummer but otherwise I'll like it as well 👍